functions/go/set-namespace/transformer/namespace.go

```diff
@@ -18,206 +18,311 @@ import (
-		oldNss = append(oldNss, `"`+oldNs+`"`)
-	}
-	msg := fmt.Sprintf("namespace %v updated to %q, %d value(s) changed", strings.Join(oldNss, ","), p.NewNamespace, count)
-	return fn.GeneralResult(msg, fn.Info)
+// SetNamespace defines structs to parse KRM resource "SetNamespace" (the custom function config) and "ConfigMap" data.
+// it provides the method "Config" to read the function configs from ResourceList.FunctionConfig
+// it provides the method "Transform" to change the "namespace" and update the "config.kubernetes.io/depends-on" annotation.
+type SetNamespace struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	NewNamespace      string `json:"namespace,omitempty"`
+	NamespaceMatcher  string `json:"namespaceMatcher,omitempty"`
 }
 
-// Config gets the attributes from different FunctionConfig formats.
-func (p *NamespaceTransformer) Config(o *fn.KubeObject) error {
+// Config gets the new namespace from FunctionConfig. It accepts three types of FunctionConfig:
+// 1. A ConfigMap object's .data.namespace
+// 2. A ConfigMap named "kptfile.kpt.dev" object's .data.name
+// 3. A SetNamespace object's .namespace
+func (p *SetNamespace) Config(o *fn.KubeObject) error {
 	switch {
 	case o.IsEmpty():
-		return fmt.Errorf("FunctionConfig is missing. Expect `ConfigMap` or `SetNamespace`")
-	case o.IsGVK("v1", "ConfigMap"):
-		p.NewNamespace = o.NestedStringOrDie("data", "namespace")
-		if p.NewNamespace == "" {
-			if o.GetName() == "kptfile.kpt.dev" {
-				p.NewNamespace = o.NestedStringOrDie("data", "name")
-				if p.NewNamespace == "" {
-					return fmt.Errorf("`data.name` should not be empty")
-				}
-			} else {
-				return fmt.Errorf("`data.namespace` should not be empty")
-			}
+		return fmt.Errorf("FunctionConfig is missing. Expect `ConfigMap(v1)` or `%s(%s/%s)`",
+			fnConfigKind, fnConfigGroup, fnConfigVersion)
+	case o.IsGVK("", "v1", "ConfigMap"):
+		var cm corev1.ConfigMap
+		o.AsOrDie(&cm)
+		p.NamespaceMatcher = cm.Data["namespaceMatcher"]
+		if cm.Data["namespace"] != "" {
+			p.NewNamespace = cm.Data["namespace"]
+			return nil
+		}
+		if cm.Data["name"] != "" && cm.Name == builtinConfigMapName {
+			p.NewNamespace = cm.Data["name"]
+			return nil
+		}
+		if p.Name == builtinConfigMapName {
+			return fmt.Errorf("`data.name` should not be empty")
 		}
-	case o.IsGVK(fnConfigAPIVersion, fnConfigKind):
-		p.NewNamespace = o.NestedStringOrDie("namespace")
+		return fmt.Errorf("`data.namespace` should not be empty")
+	case o.IsGVK(fnConfigGroup, fnConfigVersion, fnConfigKind):
+		o.AsOrDie(&p)
 		if p.NewNamespace == "" {
 			return fmt.Errorf("`namespace` should not be empty")
 		}
 	default:
-		return fmt.Errorf("unknown functionConfig Kind=%v ApiVersion=%v, expect `%v` or `ConfigMap`",
-			o.GetKind(), o.GetAPIVersion(), fnConfigKind)
+		return fmt.Errorf("unknown functionConfig Kind=%v ApiVersion=%v, expect `ConfigMap(v1)` or `%s(%s/%s)`",
```

Comment:
`Kind(group/version)` doesn't seem to be a common way to format GVK. The formatting used in kubectl is `kind.version.group`. e.g. you can run `kubectl get deployments.v1.apps`.

Reply:
ah, like the reversed domain. Thanks for the suggestion. updated.